pkg/blockbuilder/tsdb.go

```diff
@@ -139,7 +139,27 @@ func (b *TSDBBuilder) Process(ctx context.Context, rec *kgo.Record, lastBlockMax
 		// and NOT the stable hashing because that's what TSDB expects. We don't need stable hashing in block builder.
 		ref, copiedLabels := app.GetRef(nonCopiedLabels, hash)
 
+		ingestCreatedTimestamp := ts.CreatedTimestamp > 0
+
 		for _, s := range ts.Samples {
+			if ingestCreatedTimestamp && ts.CreatedTimestamp < s.TimestampMs &&
+				(processEverything || !recordAlreadyProcessed || ts.CreatedTimestamp >= lastBlockMax) &&
+				(processEverything || ts.CreatedTimestamp < blockMax) &&
+				(!nativeHistogramsIngestionEnabled || len(ts.Histograms) == 0 || ts.Histograms[0].Timestamp >= s.TimestampMs) {
```

Comment:
Sorry, I didn't know you were working on expanding the tsdb parts at the moment. I've recently merged a refactoring to the blockbuilder, that got rig of "legacy" `blockMax`, `processEverything`, etc. But now you have to resolve the conflicts here.

Let me know, if I can share any extra details from why/what was changed in https://github.com/grafana/mimir/pull/11987

Reply:
I've simplified the logic and test. I did leave some of the structure of the builder test because it gives pretty good coverage of case and I've done a lot of work to have it :)